File: codes/contest/leetcode/search-in-rotated-sorted-array.cpp
/* coding:utf-8
 * Copyright (C) dirlt
 */

#include <vector>
using namespace std;

class Solution {
   public:
    int search(vector<int>& nums, int target) {
        int size = nums.size();
        int s = 0, e = size - 1;
        while (s <= e) {
            int m = (s + e) / 2;
            if (nums[m] == target) return m;
            if (target > nums[m]) {
                if (nums[m] >= nums[s]) {
                    s = m + 1;
                } else if (target >= nums[s]) {
                    e = m - 1;
                } else {
                    s = m + 1;
                }
            } else {
                if (nums[m] <= nums[e]) {
                    e = m - 1;
                } else if (target >= nums[s]) {
                    e = m - 1;
                } else {
                    s = m + 1;
                }
            }
        }
        return -1;
    }
};

File: codes/contest/leetcode/search-in-rotated-sorted-array.py
#!/usr/bin/env python
# coding:utf-8
# Copyright (C) dirlt

class Solution(object):
    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """

        n = len(nums)
        s, e = 0, n - 1
        while s <= e:
            m = (s + e) // 2
            if nums[m] == target:
                return m
            if nums[m] >= nums[s]:
                if nums[s] <= target < nums[m]:
                    e = m - 1
                else:
                    s = m + 1
            else:
                if nums[m] < target <= nums[e]:
                    s = m + 1
                else:
                    e = m - 1
        return -1


if __name__ == '__main__':
    s = Solution()
    print(s.search([6, 7, 1, 2, 3, 4, 5], 6))
